crates/turborepo-lib/src/config/turbo_config.rs

```diff
@@ -0,0 +1,581 @@
+        let k = k.to_string();
+        if let Some(value) = environment.get(&k) {
+            if !value.is_empty() {
+                output_map.insert(mapped.to_string(), value.to_string());
+            }
+        }
+    });
+
+    let output = ConfigurationOptions {
+        api_url: None,
+        login_url: None,
+        team_slug: None,
+        team_id: output_map.get("team_id").cloned(),
+        token: output_map.get("token").cloned(),
+
+        signature: None,
+        preflight: None,
+        enabled: None,
+        timeout: None,
+    };
+
+    Ok(output)
+}
+
+impl TurborepoConfigBuilder {
+    pub fn new(base: &CommandBase) -> Self {
+        Self {
+            repo_root: base.repo_root.to_owned(),
+            override_config: Default::default(),
+            #[cfg(test)]
+            global_config_path: base.global_config_path.clone(),
+            #[cfg(test)]
+            environment: Default::default(),
+        }
+    }
+
+    // Getting all of the paths.
+    fn global_config_path(&self) -> Result<AbsoluteSystemPathBuf, ConfigError> {
+        #[cfg(test)]
+        if let Some(global_config_path) = self.global_config_path.clone() {
+            return Ok(global_config_path);
+        }
+
+        let config_dir = config_dir().ok_or(ConfigError::NoGlobalConfigPath)?;
+        let global_config_path = config_dir.join("turborepo").join("config.json");
+        AbsoluteSystemPathBuf::try_from(global_config_path).map_err(|e| ConfigError::PathError(e))
+    }
+    fn local_config_path(&self) -> AbsoluteSystemPathBuf {
+        self.repo_root.join_components(&[".turbo", "config.json"])
+    }
+    #[allow(dead_code)]
+    fn root_package_json_path(&self) -> AbsoluteSystemPathBuf {
+        self.repo_root.join_component("package.json")
+    }
+    #[allow(dead_code)]
+    fn root_turbo_json_path(&self) -> AbsoluteSystemPathBuf {
+        self.repo_root.join_component("turbo.json")
+    }
+
+    #[cfg(test)]
+    fn get_environment(&self) -> HashMap<String, String> {
+        self.environment.clone()
+    }
+
+    #[cfg(not(test))]
+    fn get_environment(&self) -> HashMap<String, String> {
+        get_lowercased_env_vars()
+    }
+
+    fn get_global_config(&self) -> Result<ConfigurationOptions, ConfigError> {
+        let global_config_path = self.global_config_path()?;
+        let mut contents = std::fs::read_to_string(&global_config_path).or_else(|e| {
+            if matches!(e.kind(), std::io::ErrorKind::NotFound) {
+                Ok(String::from(""))
+            } else {
+                Err(anyhow!(
+                    "Encountered an IO error while attempting to read {}: {}",
+                    &global_config_path,
+                    e
+                ))
+            }
+        })?;
+        if contents.is_empty() {
+            contents = String::from("{}");
+        }
+        let global_config: ConfigurationOptions =
+            serde_json::from_str(&contents).map_err(|e| ConfigError::SerdeJson(e))?;
```

Comment:
Can we add `#[from]` on the `SerdeJson` variant instead?

Reply:
It's already there, I just didn't know that worked. 😅